### PR TITLE
[FS-1030] - Anonymous records

### PR DESCRIPTION
**Pull request by dsyme**
This adds anonymous records to the F# compiler per [RFC FS-1030](https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1030-anonymous-records.md).

Please discuss on [the RFC](https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1030-anonymous-records.md) or below - if the discussion gets long we will move to the RFC discussion thread.

The implementation is not yet complete but I believe it to be stable and usable - there are just some corner cases to finish off and considerably more testing to add, specifically

Design:
* [ ] Resolve issues around field ordering (currently field ordering is significant)

Implementation:
* [x] Equality, hash and comparison on ``{| X = 1; Y = 2 |}`` Kind B types 
* [x] ToString on ``{| X = 1; Y = 2 |}`` Kind B types 
* [x] ``sprintf "%A"`` on ``{| X = 1; Y = 2 |}`` Kind B types 
* [x] Quick info 
* [x] Auto complete
* [x] Highlight symbols
* [ ] F# Interactive multiple fragments
* [ ] Field ordering 
* [ ] Find all references 
* [ ] Go-to-definition

Testing:
* [ ] Unit testing for language service features 
* [ ] LINQ queries 
* [ ] Test suitability of ``{| X = 1; Y = 2 |}`` types for working with known common libraries accepting C# anonymous types
* [ ] Add tests for negative case error messages

If anyone wants to give a hand with the testing that would be great. I think it's really important that someone else besides myself do serious testing on this feature. I'd love someone to try this out and give early feedback on usability and "fit and polish".

Note: F# Interactive currently doesn't work properly:

```
> {|  X = 1 |};;
val it : {|X : int|} = {X = 1;}

> {|  X = 1 |};;

error FS0193: internal error: Duplicate type name within an assembly.
```

### CODE REVIEW

**Comment by et1975** [comment]
I didn't see an example of decomposition, will this be possible?
```fsharp
let {| x = x |} = anonymousRecord
```
What would the Kind B syntax be like for this?
How could one distinguish between Kind A and Kind B types of same shape?

**Comment by dsyme** [comment]
@et1975 I'm not certain we need to allow decomposition for these (I'm not a big fan of pattern matching on records).  But if we did, it would be

    let {| x = x |} = anonymousRecordKindA

    let (new {| x = x |}) = anonymousRecordKindB

**Comment by vasily-kirichenko** [comment]
It's fantastic. I didn't even know it's coming or even RFCed.

1. Proper tooltip on value as a whole

image

2. No completion:

image

3. No tooltip/find usages/etc. on fields neither on definition nor on usage site (I guess there's no FSharpSymbol?)

4. %A formatting is same as for tuples. Maybe it's better to adopt the one used for records? 
```fsharp
let x = {| IntF = 1; StringF = "foo" |}
printfn "x = %A" x

> x = (1, "foo")
```

**Comment by vasily-kirichenko** [comment]
About symbol absence, I think we just should notify the sink about fields, currently there is no ones:

image

**Comment by yawaramin** [comment]
Wow! Anonymous records--very cool. Out of curiosity, are the following two representations significantly different? I.e. one is heap and the other is stack etc.?

```fsharp
type t1 = { x : int; y : string }
type t2 = {| x : int ; y : string }
```

**Comment by dsyme** [comment]
@yawaramin actually please ask that one on the RFC thread

**Comment by yawaramin** [comment]
@dsyme sorry, I deleted my question after I thought about it some more--I think I understand the difference, but I will repost on the RFC thread (i.e. https://github.com/fsharp/fslang-design/issues/170 ) in case the discussion brings value.

**Comment by dsyme** [comment]
I've updated the PR to include code for equality, comparison, ToString and record-reflection

**Comment by dsyme** [comment]
@vasily-kirichenko 

> %A formatting is same as for tuples. Maybe it's better to adopt the one used for records?

This is fixed

> No completion:

> No tooltip/find usages/etc. on fields neither on definition nor on usage site (I guess there's no FSharpSymbol?)

These should now be fixed - see the last few commits - but again I haven't tested it and may have missed something. If you have time please give it a spin again?

**Comment by dsyme** [comment]
@vasily-kirichenko There is an open question about what goto-definition should do.  At the moment the internal representations carry no range information for the fields, so there is no range to go to.

We could change that - and probably should - but it's a bit invasive and I think I'll leave it out of this particular PR

**Comment by vasily-kirichenko** [comment]
> These should now be fixed - see the last few commits - but again I haven't tested it and may have missed something. If you have time please give it a spin again?

Still nothing work: no tooltips on fields, no completion after dot.

**Comment by dsyme** [comment]
@vasily-kirichenko This is actually getting quite close to being ready.  More testing will be needed, but it's feeling good

**Comment by dsyme** [comment]
> Still nothing works: no tooltips on fields, no completion after dot.

Thanks for trying :) Will add unit tests and work up from there

**Comment by dsyme** [comment]
@vasily-kirichenko The build of the VS bits was busted due to an unused variable - did you fix that when building? thanks

**Comment by vasily-kirichenko** [comment]
Hmmm. There were no errors... Let me try again. Sorry.

**Comment by vasily-kirichenko** [comment]
@dsyme some success:



No symbol for field (so no highlighting and no find usages), no completion, but tooltips work. No idea why.

**Comment by dsyme** [comment]
@vasily-kirichenko Thanks - yes, just trying it myself, there are problems.  I'll let you know. Tooltips working for me too

**Comment by vasily-kirichenko** [comment]
FSharpSymbolUse looks OK

image

but get symbol uses returns empty array 

image

**Comment by dsyme** [comment]
@vasily-kirichenko OK symbol highlighting and completion are now working

Find all references to a symbol is not working, I think because of the problem I mentioned where ``symbol.DeclarationLocation``  and ``symbol.ImplementationLocation`` are both returning None.   These are odd symbols, because they don't have one specific declaration location (though very often there will be only one)

Inline rename will not work for similar reasons.  We should probably just have the compiler choose one occurrence as the declaration location.

**Comment by smoothdeveloper** [comment]
@dsyme the first encountered location in file compilation order (although I guess it is more complicated than that) would be the most natural place.

Nice to see such rapid progress :)

**Comment by vasily-kirichenko** [comment]
@dsyme I believe giving anon records declaration / implementation locations are crucial to get a "big picture" where a record is used in entire project.

> These are odd symbols, because they don't have one specific declaration location (though very often there will be only one)
> Inline rename will not work for similar reasons. We should probably just have the compiler choose one occurrence as the declaration location.

We can return multiple locations for Go to definition, in which case you are not jump to the file, as it is in case of single location, but all of them are shown in a special tool window (actually we did this for even single location, then fixed it so it would not work for multiple locs, but it's easy to fix).

Find usages... To make it work properly we must connect each "definition" with all its usages, which is impossible I guess? If so, we could show the very first definition and connect all usages to it, as @smoothdeveloper wrote.

Inline rename. I'm not sure if it would work ootb, but the right behavior should be to rename a field name everywhere - all "definitions" and all usages.

**Comment by dsyme** [comment]
Updated to latest master

**Comment by vasily-kirichenko** [comment]
It's aged 8 months. @dsyme any plans to finally merge this exciting PR?

**Comment by dsyme** [comment]
> @dsyme any plans to finally merge this exciting PR?

We're gradually iterating towards a viable set of PRs for F# language version 4.2, but it will take time.  This PR is certain to be part of that version, and I'm hopeful that most of [these](https://github.com/Microsoft/visualfsharp/pulls?q=is%3Apr+is%3Aopen+label%3A%22Candidate+for+F%23+4.2%22) will too, plus some additional features.

**Comment by vasily-kirichenko** [comment]
@dsyme Awesome, thanks :)

**Comment by dsyme** [comment]
Closed as targeted dev15.6 for some reason
